### PR TITLE
Initial import fo Eclipse 4DIAC IDE

### PR DESCRIPTION
**Pull request by ctron**

### CODE REVIEW

File: org.eclipse.iot.fourdiac.Ide.json [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+                "mkdir -p /app/share/icons/hicolor/{128x128,256x256}/apps",
+                "install 4diac-ide.desktop /app/share/applications/org.eclipse.iot.fourdiac.Ide.desktop",
+                "for i in 128 256; do install 4diac-ide-$i.png /app/share/icons/hicolor/${i}x${i}/apps/org.eclipse.iot.fourdiac.Ide.png; done",
+                "cp -av 4diac-ide /app",
+                "ln -s /app/jre /app/4diac-ide/jre",
+                "echo '--add-modules=ALL-SYSTEM' >> /app/4diac-ide/4diac-ide.ini"
+            ]
+        }
+    ],
+    "finish-args": [
```
Comment (nedrichards) [inline org.eclipse.iot.fourdiac.Ide.json]: on flathub we put the finish args near the top of the file just before the modules section for consistency.

File: org.eclipse.iot.fourdiac.Ide.json [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        "--env=DCONF_USER_CONFIG_DIR=.config/dconf",
+        "--env=JAVA_HOME=/app/jre",
+        "--device=dri",
+        "--share=ipc",
+        "--socket=x11",
+        "--socket=wayland",
```
Comment (nedrichards) [inline org.eclipse.iot.fourdiac.Ide.json]: do java apps actually work on wayland proper? if you just want xwayland then you shouldn't include this permission, it'll "just work"

File: org.eclipse.iot.fourdiac.Ide.json [hunk 3]
```diff
@@ -0,0 +1,62 @@
+{
+    "app-id": "org.eclipse.iot.fourdiac.Ide",
+    "runtime": "org.gnome.Platform",
+    "runtime-version": "3.28",
+    "sdk": "org.gnome.Sdk",
```
Comment (nedrichards) [inline org.eclipse.iot.fourdiac.Ide.json]: which bits of GNOME does this want that aren't in the FDo runtime (which certainly has GTK already for instance)?

File: org.eclipse.iot.fourdiac.Ide.json [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    ],
+    "command": "/app/4diac-ide/4diac-ide",
+    "modules": [
+        {
+            "name": "openjdk",
+            "buildsystem": "simple",
+            "build-commands": [ "/usr/lib/sdk/openjdk10/install.sh" ]
+        },
+        {
+            "name": "4diac-ide",
```
Comment (nedrichards) [inline org.eclipse.iot.fourdiac.Ide.json]: can this not be built from source? Or is there a specific reason you need to use an upstream binary?

**Comment by nedrichards** [comment]
bot, build org.eclipse.iot.fourdiac.Ide

**Comment by nedrichards** [comment]
Out of interest - how does Eclipse feel about Java 10? I know that Java 11 is more LTS but I haven't seen anyone look at getting that into Flathub, is that something that you, or others like @mbooth101 would be keen on?

File: org.eclipse.iot.fourdiac.Ide.json [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        "--env=DCONF_USER_CONFIG_DIR=.config/dconf",
+        "--env=JAVA_HOME=/app/jre",
+        "--device=dri",
+        "--share=ipc",
+        "--socket=x11",
+        "--socket=wayland",
+        "--share=network",
+        "--filesystem=host",
+        "--filesystem=xdg-run/dconf",
+        "--persist=/app/"
```
[the remark this answers is not in the document]
Comment (ctron) [reply org.eclipse.iot.fourdiac.Ide.json]: I am not sure which line exactly you mean.

This application can connect to its runtime on other devices, so my assumption was that this requires `--share=network`. Also does it need to work with file on the user's filesystem (thus `--filesystem=host`). My understanding was that `xdg-run/dcon` is necessary for interacting with the Gnome session. And the application will store some files during runtime in the application's main directory.

File: org.eclipse.iot.fourdiac.Ide.json (hunk 2 above)
Comment (ctron) [reply org.eclipse.iot.fourdiac.Ide.json]: This is based on Eclipse Platform and uses SWT, which is based on GTK. So if GTK supports Wayland, then this application also uses Wayland.

File: org.eclipse.iot.fourdiac.Ide.json (hunk 4 above)
Comment (ctron) [reply org.eclipse.iot.fourdiac.Ide.json]: In theory yes. But this is the official binary, signed with certificates from the Eclipse Foundation, which cannot be signed outside Eclipse infrastructure.

Aside from that I would need to replicate a good portion of the Eclipse IDE build, which I honestly am not capable of.

**Comment by ctron** [comment]
> Out of interest - how does Eclipse feel about Java 10? I know that Java 11 is more LTS but I haven't seen anyone look at getting that into Flathub, is that something that you, or others like @mbooth101 would be keen on?

I would prefer to use either Java 8 or Java 11. Java 8 would be optimal for this package. I only went with 10 as this the most recent one available out of the box.

I think it would be great to have Java 11, and maybe even base packages for OpenJ9, in addition to OpenJDK.

**Comment by ctron** [comment]
Ok, I just pushed an update to this PR. The location of `finish-args` should be fixed now. I also switched to `org.freedesktop.Sdk` and `org.freedesktop.Platform`.

**Comment by mbooth101** [comment]
Hi @nedrichards yes I am looking into getting Java 11 into Flathub. :-)

**Comment by ctron** [comment]
@nedrichards is that change ok now?

**Comment by mbooth101** [comment]
Hi I submitted a PR for Java 11, see #669

**Comment by nedrichards** [comment]
bot, build org.eclipse.iot.fourdiac.Ide

**Comment by nedrichards** [comment]
@ctron I'd prefer to hold until we get Java 11 in, which shouldn't take long - if that makes sense to you.

File: org.eclipse.iot.fourdiac.Ide.json (hunk 4 above)
Comment (nedrichards) [reply org.eclipse.iot.fourdiac.Ide.json]: the official binary thing makes sense. In this case I'd like to see that noted in the appdata description as well (see apps like VS Code which repackage upstream binaries)

File: org.eclipse.iot.fourdiac.Ide.json [hunk 6]
```diff
@@ -0,0 +1,62 @@
+{
+    "app-id": "org.eclipse.iot.fourdiac.Ide",
+    "runtime": "org.freedesktop.Platform",
+    "runtime-version": "1.6",
```
Comment (nedrichards) [inline org.eclipse.iot.fourdiac.Ide.json]: This should change to the latest release, 18.08 which now has an OpenJDK10 release and is the only runtime that OpenJDK 11 will be available for.

**Comment by nedrichards** [comment]
@ctron what's the plugin story for this app? how (if at all) do users normally extend it?

**Comment by ctron** [comment]
> @ctron what's the plugin story for this app? how (if at all) do users normally extend it?

Honestly, there is none. And I am not really sure it makes sense. 4DIAC is more like a "ready to run" application, and less an extensible platform.

As it is based on the Eclipse Platform, and the software installer indeed is present in the distribution, users could install additional plugins if they manually configure P2 repos.

So I just gave it a try and it did install an additional plugin. It even survived a "flatpak udpdate".

File: org.eclipse.iot.fourdiac.Ide.json (hunk 6 above)
Comment (ctron) [reply org.eclipse.iot.fourdiac.Ide.json]: Maybe someone should update the docs ;-) I updated to `18.08`, that looks good.

**Comment by ctron** [comment]
> @ctron I'd prefer to hold until we get Java 11 in, which shouldn't take long - if that makes sense to you.

If that means today or tomorrow, then yes. I am not sure if Java 11 will work, as the application is not targeted for versions beyond Java 8. It only works on Java 9+ with some flags. But I think it is work a try.

File: org.eclipse.iot.fourdiac.Ide.json (hunk 4 above)
Comment (ctron) [reply org.eclipse.iot.fourdiac.Ide.json]: Ok, I add a paragraph about that.

**Comment by ctron** [comment]
I just pushed a new change, this should limit the architecture to "x86_64" and also copies the "appdata" file to the right location.

**Comment by nedrichards** [comment]
OK - sounds good on the plugin thing, I must've been thinking of the main eclipse platform.

**Comment by nedrichards** [comment]
bot, build org.eclipse.iot.fourdiac.Ide

**Comment by ctron** [comment]
It looks like the Java 11 PR takes a bit longer. Would it be ok to continue with the tested Java 10 dependency and go for Java 11 once this is sorted out?

**Comment by mbooth101** [comment]
FWIW TingPing just provided a positive second review for the jdk11 PR -- so maaaaybe it can be in today? Nudge nudge, wink wink ;-)

**Comment by nedrichards** [comment]
Java 11 just merged and building now.

**Comment by ctron** [comment]
It still looks like there is some kind of problem with that build, as OpenJDK 11 cannot be found.

As everything else in this PR seems to be ready, I was hoping to get this merged and be able to present this on Thursday at a conference. I know this is not your problem but mine, but still it think "is available on flathub" sounds better than "may be on flathub in the future".

So it would be great if someone could either have a look at the issues with OpenJDK 11, or if we could go ahead with this current PR.

**Comment by mbooth101** [comment]
Another build in-progress: https://flathub.org/builds/#/builders/2/builds/5336

**Comment by ctron** [comment]
> Another build in-progress: https://flathub.org/builds/#/builders/2/builds/5336

From what I saw, it is broken again …

**Comment by mbooth101** [comment]
Third time's the charm: https://flathub.org/builds/#/builders/2/builds/5340

**Comment by ctron** [comment]
So, unfortunately Java 11 doesn't work. The Eclipse platform version by this version of 4DIAC doesn't isn't compatible with Java 11 and simply doesn't start. There is nothing that I can do in order to work around this.

So either it is Java 10, or I have to manually package Java 8, which I would like to avoid.

**Comment by mbooth101** [comment]
Ah, so this app is based on an old version of the Eclipse Platform then :-(

If it's at least based on Eclipse 4.7.3a then Java 10 should work no problem.

**Comment by ctron** [comment]
It is `4.7.3.v20180330-0640`, which works when adding `--add-modules=ALL-SYSTEM`. But it looks like Java 11 is doing something different here, as this no longer works.

**Comment by nedrichards** [comment]
OK - will add this now then and we can look at Java 11 in conjunction with upstream!

**Comment by nedrichards** [comment]
Repository has been created: https://github.com/flathub/org.eclipse.iot.fourdiac.Ide

**Comment by ctron** [comment]
Thanks for taking care of this!

**Pull request closed without merging**